### PR TITLE
🕊️ DPO padding free

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

demo; further experiments in next comment

```python
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl import DPOConfig, DPOTrainer
import torch

model_id = "Qwen/Qwen2-0.5B-Instruct"
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16, attn_implementation="flash_attention_2")
tokenizer = AutoTokenizer.from_pretrained(model_id)
dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train[:10%]")
training_args = DPOConfig(output_dir="Gemma2-2B-DPO-pf", max_prompt_length=128, max_completion_length=128, logging_steps=10, padding_free=True)
trainer = DPOTrainer(model=model, args=training_args, train_dataset=dataset, processing_class=tokenizer)
trainer.train()
```

With and without padding-free (not sure why they don't match exactly, the logits do precisely match though)
Screenshot 2024-12-26 at 22 00 06

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -89,6 +89,7 @@ jobs:
           python -m pip install -U git+https://github.com/huggingface/datasets.git
           python -m pip install -U git+https://github.com/huggingface/transformers.git
           python -m pip install ".[dev]"
+
```
Comment (qgallouedec) [inline .github/workflows/tests.yml:92]:
```suggestion
```

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -1149,12 +1158,22 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
                 num_logits_to_keep = loss_mask.shape[1] - first_compute_index
                 model_kwargs["num_logits_to_keep"] = num_logits_to_keep.item() + 1  # +1 for the first label
 
-            outputs = model(input_ids=input_ids, attention_mask=attention_mask, **model_kwargs)
+            if self.padding_free:
+                # Flatten the input_ids, position_ids, and loss_mask
+                # input_ids = [[a, b, c, 0], ->     input_ids = [a, b, c, d, e, f, g]
+                #              [d, e, f, g]]     position_ids = [0, 1, 2, 0, 1, 2, 3]
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1164]:
```suggestion
                # Flatten the input_ids, position_ids, and loss_mask
                # input_ids = [[a, b, c, 0], ->     input_ids = [[a, b, c, d, e, f, g]]
                #              [d, e, f, g]]     position_ids = [[0, 1, 2, 0, 1, 2, 3]]
```

File: trl/trainer/dpo_config.py [hunk 3]
```diff
@@ -145,6 +145,9 @@ class DPOConfig(TrainingArguments):
             useful for saving memory and speeding up training by not computing the logits for all tokens, especially in
             scenarios when working with very long prompts where labels are ignored (-100).
             [Read more](https://huggingface.co/docs/transformers/main/model_doc/llama#transformers.LlamaForCausalLM)
+        padding_free (`bool`, *optional*, defaults to `False`):
+            Whether the forward passes are performed without padding, i.e. flattening all the samples in the batch
+            into a single sample, associated with a position_ids vector. Only possible with flash-attention.
```
Comment (qgallouedec) [inline trl/trainer/dpo_config.py:150]:
```suggestion
        padding_free (`bool`, *optional*, defaults to `False`):
            Whether forward passes are performed without padding by flattening all sequences in the batch 
            into a single continuous sequence. This approach requires associating a `position_ids` vector to track 
            positional information. Currently, this is only supported with the `flash_attention_2` mechanism, as it 
            can handle the flattened batch structure.
```

File: trl/trainer/dpo_trainer.py [hunk 4]
```diff
@@ -1152,12 +1164,22 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
...
                 model_kwargs["num_logits_to_keep"] = num_logits_to_keep.item() + 1  # +1 for the first label
 
-            outputs = model(input_ids=input_ids, attention_mask=attention_mask, **model_kwargs)
+            if self.padding_free:
+                # Flatten the input_ids, position_ids, and loss_mask
+                # input_ids = [[a, b, c, 0], ->     input_ids = [[a, b, c, d, e, f, g]]
+                #              [d, e, f, g]]     position_ids = [[0, 1, 2, 0, 1, 2, 3]]
+                input_ids = input_ids[attention_mask.bool()].unsqueeze(0)
+                loss_mask = loss_mask[attention_mask.bool()].unsqueeze(0)
+                model_kwargs["position_ids"] = attention_mask.cumsum(1)[attention_mask.bool()].unsqueeze(0) - 1
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1173]:
```suggestion
                position_ids = attention_mask.cumsum(1)[attention_mask.bool()].unsqueeze(0) - 1
                model_kwargs["position_ids"] = position_ids
```

File: trl/trainer/dpo_trainer.py [hunk 5]
```diff
@@ -1208,8 +1241,22 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
...
-        output["mean_chosen_logits"] = logits[:num_examples][loss_mask[:num_examples]].mean()
-        output["mean_rejected_logits"] = logits[num_examples:][loss_mask[num_examples:]].mean()
+
+        # Compute the mean logits
+        if self.padding_free:
+            # pos_id contains a sequence of range identifiers (e.g., [[0, 1, 2, 0, 1, 2, 3, ...]]).
+            # There are 2*num_examples ranges in total: the first half corresponds to the chosen tokens,
+            # and the second half to the rejected tokens.
+            # To find the start of the rejected tokens, we look for the num_examples+1-th zero in pos_id.
+            split_idx = (model_kwargs["position_ids"] == 0).nonzero(as_tuple=True)[1][num_examples]
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1251]:
```suggestion
            split_idx = (position_ids == 0).nonzero(as_tuple=True)[1][num_examples]
```

**Comment by qgallouedec** [comment]
## Regression test:

```python
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl import DPOConfig, DPOTrainer
import torch

model_id = "Qwen/Qwen2-0.5B-Instruct"
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16, attn_implementation="flash_attention_2")
tokenizer = AutoTokenizer.from_pretrained(model_id)
dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train[:10%]")
# dataset = load_dataset("trl-internal-testing/zen", "standard_preference", split="train")
training_args = DPOConfig(output_dir="Qwen2-0.5B-DPO-no_pf", max_prompt_length=128, max_completion_length=128, logging_steps=10, padding_free=False)
trainer = DPOTrainer(model=model, args=training_args, train_dataset=dataset, processing_class=tokenizer)
trainer.train()
```

Is the new `padding_free=False` (`no_pf` in screenshot) equivalent to DPO on current main branch (`main` in screenshot)? -> yes

Screenshot 2025-01-07 at 20 15 31

Does `padding_free=True` (`pf` in screenshot) results match `padding_free=False` (`no_pf` in screenshot) results? -> Yes

Screenshot 2025-01-07 at 20 19 41

(note: on screenshots its written "Gemma" but it's actually a Qwen model trained)

File: trl/trainer/dpo_trainer.py [hunk 6]
```diff
@@ -1149,15 +1161,26 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
...
+                model_kwargs["attention_mask"] = attention_mask
 
-            outputs = model(input_ids=input_ids, attention_mask=attention_mask, **model_kwargs)
+            outputs = model(input_ids, **model_kwargs)
+            logits = outputs.logits
 
             # Offset the logits by one to align with the labels
-            logits = outputs.logits[:, :-1, :]
-            labels = input_ids[:, 1:].clone()
-            loss_mask = loss_mask[:, 1:].bool()
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1160]: rolling works in both cases (flattened tensors and batched)

```python
# input_ids = [[1, 2, 3, 4],
#              [5, 6, 7, 8]]
# loss_mask = [[0, 0, 1, 1],
#              [0, 0, 0, 1]]
labels = input_ids[:, 1:].clone()
loss_mask = loss_mask[:, 1:].bool()
# labels =    [[2, 3, 4],
#              [6, 7, 8]]
# loss_mask = [[0, 1, 1],
#              [0, 0, 1]]

# But

# input_ids = [[1, 2, 3, 4, 5, 6, 7, 8]]
# loss_mask = [[0, 0, 1, 1, 0, 0, 0, 1]]
labels = input_ids[:, 1:].clone()
loss_mask = loss_mask[:, 1:].bool()
# labels =    [[2, 3, 4, 5, 6, 7, 8]]
# loss_mask = [[0, 1, 1, 0, 0, 0, 1]]
```

The first token of the first sequence (1) is removed but not the first token of the second sequence (5). To align the labels while keeping a consistent behaviour across sequences in the batch, we use roll instead. The only difference is that the first token, instead of being discarded, is appended to the end.

**Review by August-murr — APPROVED** [review]
[LGTM!](https://youtu.be/rR4n-0KYeKQ?t=30)
